### PR TITLE
Fix #10863: Update API test for fluent API chaining

### LINKED ISSUE #10863: Update API test for fluent API chaining
To avoid issues like #10808

### PR DESCRIPTION
**Pull request by anpete**

### CODE REVIEW

File: src/EFCore.SqlServer/Extensions/SqlServerDbContextOptionsBuilderExtensions.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ namespace Microsoft.EntityFrameworkCore
     /// <summary>
     ///     SQL Server specific extension methods for <see cref="DbContextOptionsBuilder" />.
     /// </summary>
-    public static class SqlServerDbContextOptionsExtensions
+    public static class SqlServerDbContextOptionsBuilderExtensions
```
Comment (anpete) [inline src/EFCore.SqlServer/Extensions/SqlServerDbContextOptionsBuilderExtensions.cs]: @ajcvickers This break OK?

Comment (smitpatel) [reply src/EFCore.SqlServer/Extensions/SqlServerDbContextOptionsBuilderExtensions.cs]: Tool did not catch it then it should be fine.

Comment (anpete) [reply src/EFCore.SqlServer/Extensions/SqlServerDbContextOptionsBuilderExtensions.cs]: Tool did :trollface:

Comment (ajcvickers) [reply src/EFCore.SqlServer/Extensions/SqlServerDbContextOptionsBuilderExtensions.cs]: Probably should not make this change--binary compat break and doesn't seem to add much value other than cleanup.

File: test/EFCore.Tests/ApiConsistencyTestBase.cs [hunk 2]
```diff
@@ -25,6 +25,26 @@ public abstract class ApiConsistencyTestBase
         protected const BindingFlags AnyInstance
             = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
 
+        protected virtual IEnumerable<Type> FluentApiTypes
+            => Enumerable.Empty<Type>();
+
+        [Fact]
+        public void Fluent_api_methods_should_not_return_void()
+        {
+            var voidMethods
```
Comment (ajcvickers) [inline test/EFCore.Tests/ApiConsistencyTestBase.cs]: Would be nice to test if a method is "called on" a generic builder, then it returns as generic builder. There are lots of places where we have generic overloads, and it's easy to miss them or get the return type wrong. But this is probably a separate issue.

**Comment by orubel** [comment]
This is not API Chaining(tm) which was created and trademarked by myself several years ago and is a process in the BeAPI Framework. What you are doing is chaining request/response to the next request/response to the next request/response, etc.

This trademark has been enforced against Mulesoft, NordicAPI, Apigee, and others.

To read on how to implement, go to the documentation page for API Chaining https://orubel.github.io/Beapi-API-Framework/chain.html

There it clearly states 'API Chaining as a process is only considered successfully implemented if:

- it uses one REQUEST/RESPONSE
- the client can dynamically create chains based on endpoints accessible by their ROLE/PRIVILEGE; - no hardcoded chains
- no more than ONE non-safe REQUEST method is used in a chain; chains CAN chain off each other however

Please read the documentation

**Comment by bricelam** [comment]
https://www.martinfowler.com/dslCatalog/methodChaining.html

**Comment by orubel** [comment]
Indeed.And method chaining is NOT 'API Chaining' because it requires a request/response with each resource/endpoint called in the chain.

API Chaining works with the new api pattern built for distributed architectures wherein the communication logic/data are abstracted from the business logic so that the request can be 'recycled' and reused for continuous calls in the change (like a batch call).

Things like Method Chaining, GraphQL all require separate request/responses for each resopurce/endpoint and thus have higher processing time, higher I/O.

A proper API Chaining implementation can reduce code and call time by 180%

https://www.slideshare.net/bobdobbes/api-pattern

Have given this talkat APIDays, APIWorld and Restfest as well as when consulting with VMWare, Netflix, Mulesoft, Mashery/Tibco, AWS API Gateway team, Nike, and others.

If you feel up to rewriting to reduce your throughput and code and solving for the architectural cross cutting concern to make API Chaining work, let me know an I'll happily advice.

But if you are using 'Method Chaining', call it Method Chaining... because it is not an API Chain

**Comment by bricelam** [comment]
You're misreading the title. It's referring to (method) chaining in the parts of our product we (loosely) classify as fluent API.

**Comment by orubel** [comment]
I understand you are using 'method chaining'; but I like to say, "Pepsi is not 'a Coke'... so don't call it a Coke". Heh

But you get the idea...  it's accurate in the sense that a trademark cannot be presented as another product. You see what I'm getting at?

BUT PLEASE, do implement the process and then use it all you want! Its a bit of work but you will see a 180% speed increase, easier synchronization of state in distributed architectures and the ability to share state.

It just requires understanding that the OLD API Pattern was invented during centralized architectures and is INTENDED for centralized architectures; you cannot use it in a distributed architecture without unsynchronized/unshared state and an architectural cross cutting concern (see https://www.slideshare.net/bobdobbes/api-pattern#15 )

You may be doing this in C# but you should still be able to implement with a HandlerInterceptor at the 'Front Controller'. That will enable you to abstract the communication layer from the business logic. You can see an example of this in Java here:  

https://github.com/orubel/Beapi-API-Framework/blob/master/grails-app/controllers/net/nosegrind/apiframework/ApiFrameworkInterceptor.groovy

That keeps your controllers from looking like this: https://www.slideshare.net/bobdobbes/api-pattern#32

... and instead looking more like this: https://www.slideshare.net/bobdobbes/api-pattern#33

Notice how most of the logic is abstracted from the controller; this is because when it is bound at the controller/business logic, you can't share the hardcoded state of an endpoint with other services nor can you synchronize it.

Best of luck and let me know if you need anything. Am always happy to help.

**Comment by orubel** [comment]
And while you state 'The concept API chaining is not mentioned--even erroneously', the title above state : "Update API test for fluent API chaining"

If you wanted to rephrase to avoid usage of the trademark, you can state 'Update API test for fluent chaining of API's'

You see? This avoids the use of the trademark and possible confusion that your product is using the process when it is not.

**Comment by Eilon** [comment]
@orubel we will get back to you ASAP regarding this.

**Comment by roji** [comment]
@orubel out of curiosity, could you please provide a link to your trademark?

Note: I'm not part of Microsoft in any way (am just an open source contributor), but I'm interested as I know of many uses of the words "API chaining" on the Internet and in programming which have nothing to do with your particular concept.

**Comment by ralmsdeveloper** [comment]
@roji  I'm also curious about this!

**Comment by orubel** [comment]
You will not find it on a TESS searchas it is a 'common law trademark' associated with a process used in a product. You can read more about it here [USPTO -
 Trademark Facts : Page 9] (https://www.uspto.gov/sites/default/files/documents/BasicFacts.pdf):

"Be aware that any searches you conduct on TESS are limited to the USPTO’s database of federal
trademark applications and registrations and do not include the marks of other parties who may
have trademark rights but no federal registration. These rights, known as “common law” rights,
are based solely on use of the mark in commerce within a particular geographic area. Common
law rights may be stronger than those based on a registration, if the common law use is earlier
than the use that supports the registration. Therefore, it is critical to learn whether superior
common law rights exist, by searching the Internet for websites and articles that reference similar
marks that are related to your goods and services. You should also search state trademark
databases and business name databases. Because searching is very complex, you should seriously
consider hiring a trademark attorney to assist you with a “full” or “comprehensive” trademark
search, as discussed below. "

**Comment by orubel** [comment]
I have this conversation ALOT (as you can see) because people FREQUENTLY tell me that I don't have a trademark and then I have to take legal action. I'm hoping can provide you with as much information up front as I possibly can.

I created the process years ago and have been writing on 'API Abstraction', the new API Pattern, shared I/O state and other concepts... all of which are required to be able to do API Chaining(tm).

Bcause if you don't do it with a single request/response (like a monad), you are merely doing method chaining (like you stated earlier).

And I try to be clear and up front that my code is open and I share the trademark but people have to implement the process in order to use the trademark (as the trademark is associated with the process).

And I also make the process open as well.

The intent is to avoid pollution of the trademark and to create education of the difference between method chaining and 'api chaining(tm)'

**Comment by orubel** [comment]
By the way if you properly implement this, you will get a near 200% speed and scale increase and reduction in code by about 80%

**Comment by ralmsdeveloper** [comment]
@orubel one question.
All this your arguments, are because of the title, or some code of yours was used?

**Comment by roji** [comment]
@orubel thanks for the response... Again, I'm asking all this out of curiosity (I don't have any product that uses ) and have no connection to Microsoft or this specific discussion.

Yeah, I'm aware of common law tradmarks. However, I do have several questions and hope you'll have the patience to answer them (which I appreciate, and I also definitely respect your technology):

1. It's my understanding that you're also claiming trademark rights on "api abstraction" (seems that way from [this thread](https://jaxenter.com/connecting-world-111477.html)). This makes me quite worried as I know I've used it many times in the past in ways not connected with you or your product in any way.
2. I can find numerous mentions of "api chaining" from a long time ago (https://dspace.mit.edu/handle/1721.1/33295 - academic article from 2005, https://www.embedthis.com/esp/doc/users/pages.html from 2011 according to Google). I may be mistaken and these may be actual references to your own concept/technology - please correct me if so - or you may be claiming to have trademarked API chaining that long ago (in which case it would be good to see a link). But otherwise wouldn't these be a prior usage ("prior art") of this term, demonstrating that it was in use long before you started using it?
3. 1. I'm a bit confused on why you're calling it a trademark - which is the source of a good - rather than a patent, which seems more appropriate for your case (being an invention or technique).

**Comment by Eilon** [comment]
Hi folks, let's please take other discussions on this topic to another forum. You can use each other's GitHub contact info to reach each other.

**Comment by roji** [comment]
Apologies @Eilon...

@orubel if you're interested in carrying on this conversation I can open an issue in one of my repos or we can do it on one of yours.

**Comment by orubel** [comment]
Trademarks can be associated with process. The only thing I'm asking is that you don't use the trademark unless you implement the process. And if you implement the process, I give full rights for use.

I'm not going to argue over whether you think its valid and past conversations. Forward to Microsoft legal if you want. I'm just following protocol.

**Pull request merged**